core/src/main/java/com/datastax/oss/driver/internal/core/context/LifecycleListener.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.context;
+
+import com.datastax.oss.driver.api.core.session.SessionBuilder;
+
+/**
+ * A component that gets notified of certain events in the session's lifecycle.
+ *
+ * <p>This is intended for third-party extensions, no built-in components implement this.
+ */
+public interface LifecycleListener extends AutoCloseable {
+
+  /**
+   * Invoked when the session is ready to process user requests.
+   *
+   * <p>This corresponds to the moment when the {@link SessionBuilder#build()} returns, or the
+   * future returned by {@link SessionBuilder#buildAsync()} completes. If the session initialization
+   * fails, this method will not get called.
+   *
+   * <p>This method is invoked on a driver thread, it should complete relatively quickly and not
+   * block.
+   */
+  void onSessionReady();
```

Comment:
We will need to get the unique session identifier as a parameter here (and for `close()` as well)
If the listener will get the notification for every new session it may want to create a resource for that specific session. Then to be able to identify that resource in the `close()` callback we would need to have session id.

Reply:
We've discussed this elsewhere already, but just to document the answer: a given context is associated to exactly one session, so that won't be an issue.